### PR TITLE
Fix some memory leaks in WinHttpRequestState

### PR DESCRIPTION
**Pull request by davidsh**
When HTTP requests are canceled, we are leaking WinHttpRequestState objects and some child objects. Since WInHttpRequestState is strongly pinned, it's important to release things as soon as possible.

This fix addresses part of the leaks involved in #8819 by removing references to some fields in the state object related to the request/response phase (CancellationToken).  In addition, this fix addresses some cases where the state object is created and pinned but the request is canceled before we hook up the state object to the WInHTTP status callback. Because of not being hooked up yet, the state object won't get HANDLE_CLOSING notifications and remains pinned forever.

We now pin the state object later when we're sure that we've hooked up the status callback.

I also added some DEBUG tracking to help diagnose the rest of the leaks of #8819.

### CODE REVIEW

**Comment by davidsh** [comment]
cc: @stephentoub @CIPop @DavidObando

**Comment by DavidObando** [comment]
Looks good to me. Thanks! :shipit:

**Comment by davidsh** [comment]
Build timeouts:
http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/windows_nt_debug_prtest/1829/
http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/windows_nt_release_prtest/1814/

**Comment by davidsh** [comment]
Test Innerloop Windows_NT Debug Build and Test please
Test Innerloop Windows_NT Release Build and Test please

**Comment by stephentoub** [comment]
LGTM

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs [hunk 1]
```diff
@@ -16,6 +16,15 @@ namespace System.Net.Http
 {
     internal sealed class WinHttpRequestState : IDisposable
     {
+#if DEBUG
+        private static int s_dbg_allocated = 0;
+        private static int s_dbg_pin = 0;
```
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs]: Should there be a Debug.Assert on these getting back to 0 at some point to detect other kinds of leaks?

**Comment by CIPop** [comment]
LGTM

I like the debug helpers. Maybe we could add some `Debug.Assert` statements to fail faster if we're leaking in other ways?

**Comment by davidsh** [comment]
>Should there be a Debug.Assert on these getting back to 0 at some point to detect other kinds of leaks?

Right now, I'm using these numbers to look at dump files.  They aren't going to zero and it's not clear what kind of test to write to measure that.  But these counters are very useful when looking at dump files.

**Comment by davidsh** [comment]
>I like the debug helpers. Maybe we could add some Debug.Assert statements to fail faster if we're leaking in other ways?

I might do that later once we figure out all root cause of the main leaks.

**Comment by davidsh** [comment]
@stephentoub Do you know why the Windows CI tests timed out after 3+ hours.  It was in the Http tests which finished ok:

>14:11:49   Finished:    System.Net.Http.Functional.Tests
14:11:49   
14:11:49   === TEST EXECUTION SUMMARY ===
14:11:49      System.Net.Http.Functional.Tests  Total: 492, Errors: 0, Failed: 0, Skipped: 20, Time: 24.823s
14:11:49   Finished running tests.  End time=14:12:42.42, Exit code = 0

**Comment by stephentoub** [comment]
> Do you know why the Windows CI tests timed out after 3+ hours.

Do you have a link to the leg that timed out?

**Comment by davidsh** [comment]
It's already in the comments I posted above.

Build timeouts:
http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/windows_nt_debug_prtest/1829/
http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/windows_nt_release_prtest/1814/

**Comment by stephentoub** [comment]
It's the System.Net.Http.WinHttpHandler.Unit.Tests tests.  You can see in the log they start and never finish.

**Comment by davidsh** [comment]
>It's the System.Net.Http.WinHttpHandler.Unit.Tests tests. You can see in the log they start and never finish.

Thanks! I figured it out. I had to adjust the unit tests to account for the new WinHttpRequestState.Pin() method.

**Comment by stephentoub** [comment]
Test Innerloop OSX Debug Build and Test please

**Pull request merged**